src/audio/eq_iir.c

```diff
@@ -143,99 +130,128 @@ static void eq_iir_free_delaylines(struct iir_state_df2t *iir)
+		trace_eq_error("enr");
+		return -EINVAL;
+	}
 
 	/* Collect index of response start positions in all_coefficients[]  */
 	j = 0;
 	assign_response = &config->data[0];
 	coef_data = &config->data[config->channels_in_config];
-	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++) {
+	trace_eq("idx");
+	for (i = 0; i < SOF_EQ_IIR_MAX_RESPONSES; i++) {
 		if (i < config->number_of_responses) {
-			response_index[i] = j;
-			j += NHEADER_DF2T
-				+ NBIQUAD_DF2T * coef_data[j];
+			eq = (struct sof_eq_iir_header_df2t *)&coef_data[j];
+			lookup[i] = eq;
+			trace_value(j);
+			j += SOF_EQ_IIR_NHEADER_DF2T
+				+ SOF_EQ_IIR_NBIQUAD_DF2T * eq->num_sections;
 		} else {
-			response_index[i] = 0;
+			lookup[i] = NULL;
 		}
 	}
 
 	/* Initialize 1st phase */
 	for (i = 0; i < nch; i++) {
-		resp = assign_response[i];
-		if (resp > config->number_of_responses - 1)
-			return -EINVAL;
+		/* Check for not reading past blob response to channel assign
+		 * map. If the blob has smaller channel map then apply for
+		 * additional channels the response that was used for the first
+		 * channel. This allows to use mono blobs to setup multi
+		 * channel equalization without stopping to an error.
+		 */
+		if (i < config->channels_in_config)
+			resp = assign_response[i];
+		else
+			resp = assign_response[0];
 
 		if (resp < 0) {
-			/* Initialize EQ channel to bypass */
+			/* Initialize EQ channel to bypass and continue with
+			 * next channel response.
+			 */
 			iir_reset_df2t(&iir[i]);
-		} else {
-			/* Initialize EQ coefficients */
-			idx = response_index[resp];
-			s = iir_init_coef_df2t(&iir[i], &coef_data[idx]);
-			if (s > 0)
-				size_sum += s;
-			else
-				return -EINVAL;
+			continue;
 		}
 
+		if (resp > config->number_of_responses - 1)
+			return -EINVAL;
+
+		/* Initialize EQ coefficients */
+		eq = lookup[resp];
+		s = iir_init_coef_df2t(&iir[i], eq);
+		if (s > 0)
+			size_sum += s;
+		else
+			return -EINVAL;
 	}
 
+	/* If all channels were set to bypass there's no need to
+	 * allocate delay. Just return with success.
+	 */
+	if (!size_sum)
+		return 0;
+
 	/* Allocate all IIR channels data in a big chunk and clear it */
 	iir_delay = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, size_sum);
-	if (iir_delay == NULL)
+	if (!iir_delay)
 		return -ENOMEM;
 
 	memset(iir_delay, 0, size_sum);
 
 	/* Initialize 2nd phase to set EQ delay lines pointers */
 	for (i = 0; i < nch; i++) {
 		resp = assign_response[i];
-		if (resp >= 0) {
+		if (resp >= 0)
 			iir_init_delay_df2t(&iir[i], &iir_delay);
-		}
-
 	}
-
 	return 0;
 }
 
 static int eq_iir_switch_response(struct iir_state_df2t iir[],
-	struct sof_eq_iir_config *config, uint32_t ch, int32_t response)
+				  struct sof_eq_iir_config *config,
+				  uint32_t ch, int32_t response)
 {
 	int ret;
 
 	/* Copy assign response from update and re-initialize EQ */
-	if ((config == NULL) || (ch >= PLATFORM_MAX_CHANNELS))
+	if (!config || ch >= PLATFORM_MAX_CHANNELS)
```

Comment:
ch > PLATFORM_MAX_CHANNELS?

Reply:
E.g. When PLATFORM_MAX_CHANNELS is 4 the OK range for ch is 0..3.

Reply:
I will rewrite this function & surroundings later (it works but is not efficient) to avoid updating all EQ channels when one channel has changed. Better to collect SOF_EQ_IIR_IDX_SWITCH data for all channels first and then call this function.